rust/src/builders/tx_builder.rs

```diff
@@ -662,25 +507,44 @@ impl TransactionBuilder {
         self.validity_start_interval = Some(*validity_start_interval)
     }
 
-    /// TODO: replace with add_cert that takes in a CertificateBuilderResult
-    pub fn set_certs(&mut self, certs: &Certificates) -> Result<(), JsError> {
-        self.certs = Some(certs.clone());
-
-        let mut vkey_set = HashSet::<Ed25519KeyHash>::new();
-        for cert in &certs.0 {
-            add_cert_vkeys(cert, &mut vkey_set)?;
-        };
-        for vkey in &vkey_set {
-            self.input_types.vkeys.insert(vkey.clone());
+    pub fn add_cert(&mut self, result: &CertificateBuilderResult) {
+        let mut certs = self.certs.clone().unwrap_or(Certificates::new());
+        certs.add(&result.cert);
+        self.certs = Some(certs);
+        if let Some(ref data) = result.aggregate_witness {
+            self.witness_set_builder.add_input_aggregate_witness_data(data);
```

Comment:
Note to self: verify these fake witnesses are only used later for fee estimation and are separate from the real witnesses we add on the build() call